packages/remix-react/components.tsx

```diff
@@ -1335,8 +1601,38 @@ type JsonPrimitives =
   | Number
   | Boolean
   | null;
+
 type NonJsonPrimitives = undefined | Function | symbol;
-type SerializeType<T> = T extends JsonPrimitives
+
+type SerializeType<T, UseDeferred = false> = UseDeferred extends false
```

Comment:
@colinhacks and @mattpocock would either of you be able to review some of this TypeScript magic? 😅

Reply:
Same network serialization type guard but with promise support:
```ts
type Test = SerializeType<{ a: Promise<Date>, b: Date }, true>;
// Test should equal Test2
type Test2 = { a: Promise<string>, b: string };
```